src/main/clojure/osgi/core.clj

```diff
@@ -1,76 +1,150 @@
-             Bundle/STOPPING    :stopping
-             Bundle/UNINSTALLED :uninstalled
-             Bundle/STARTING    :starting})
+(defn- methods-of
+  "Attempts to use Java's reflection facilities to get
+  string representations of function signatures. Takes
+  any Java object."
+  [x]
+  (->> (.getMethods (class x))
+       (map #(.toGenericString %))))
 
-(defn dict->map [dict]
+(defn- dict->map
+  [dict]
   (into {}
         (map vector
              (map (comp keyword str/lower-case)
                   (Collections/list (. dict keys)))
              (Collections/list (.elements dict)))))
 
-(defn bundle->map [bundle]
+(defn- pair
+  [p]
+  (let [k (keyword (.getName p)) v (.getValue p)]
+    [k
+     (case k
+       (:uses :bundle-classpath :embed-dependency)
+       (str/split v #",")
+       v)]))
+
+(defn- clause->map
+  [clause]
+  [(.getName clause)
+   (into {} (map pair
+                 (concat (.getDirectives clause)
+                         (.getAttributes clause))))])
+
+(defn- parse-header
+  [headers k]
+  (let [header (get headers k)]
+    (assoc headers k (into {} (map clause->map (Parser/parseHeader header))))))
+
+(defn- parse-headers
+  [headers]
+  (-> headers
+      (parse-header :export-package)
+      (parse-header :export-service)
+      (parse-header :import-package)
+      (parse-header :require-capability)
+      (parse-header :embedded-artifacts)))
+
+(defn- bundle->map
+  [bundle]
   {:id       (.getBundleId bundle)
    :name     (.getSymbolicName bundle)
    :version  (.getVersion bundle)
-   :headers  (-> (.getHeaders bundle) dict->map)
+   :headers  (-> (.getHeaders bundle) dict->map parse-headers)
    :location (.getLocation bundle)
-   :state    (get states (.getState bundle) :unknown)
+   :state    (get bundle-states (.getState bundle) :unknown)
    :modified (.getLastModified bundle)
    :fragment (.isFragment bundle)})
 
+; Public operational API
 (defn start! [bundle] (.start bundle))
 (defn stop! [bundle] (.stop bundle))
 (defn uninstall! [bundle] (.uninstall bundle))
+(defn shutdown! [] (-> (bundle-cxt) (. getBundle 0) (. stop)))
 
-(defn bundle-cxt []
-  (.. FrameworkUtil (getBundle Nrepl) getBundleContext))
-
-(defn shutdown! []
-  (-> (bundle-cxt) (. getBundle 0) (. stop)))
-
-(defn bundle-list []
-  (.. (bundle-cxt) getBundles))
-
-(defn bundle-by-location [coor]
-  (->> (bundle-list) (filter #(-> % (.. getLocation) (= coor))) first))
-
+; Services
 (defn get-services [name]
   (let [ctx (bundle-cxt)]
     (->> (. ctx (getAllServiceReferences name nil))
          (map #(. ctx (getService %))))))
 
+; Features
 (defn get-features-service []
   (-> "org.apache.karaf.features.FeaturesService" get-services first))
+(defn list-repositories [] (.listRepositories (get-features-service)))
+(defn list-features [] (->> (list-repositories) (mapcat #(.getFeatures %))))
 
-(defn list-repositories []
-  (.listRepositories (get-features-service)))
+(defn- bundle-list
```

Comment:
Do you not want people to access the raw osgi bundles?

Reply:
Their choice. I just did this for now because I'm only interested in datified versions of the bundles.